File: src/main/java/com/hanocybous/ecommercesystem/dto/product/ProductDto.java
package com.hanocybous.ecommercesystem.dto;

import java.util.Objects;

public record ProductDto(
        Long id,
        String name,
        double price,
        String category,
        int quantity
) {

    public ProductDto {
        Objects.requireNonNull(id);
        Objects.requireNonNull(name);
        Objects.requireNonNull(price);
        Objects.requireNonNull(category);
        Objects.requireNonNull(quantity);

        if (name.isBlank()) {
            throw new IllegalArgumentException("Name cannot be blank");
        }
        if (price < 0) {
            throw new IllegalArgumentException("Price cannot be negative");
        }
        if (category.isBlank()) {
            throw new IllegalArgumentException("Category cannot be blank");
        }
        if (quantity < 0) {
            throw new IllegalArgumentException("Quantity cannot be negative");
        }

    }

    @Override
    public String toString() {
        return "ProductDto{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", price=" + price +
                ", category='" + category + '\'' +
                ", quantity=" + quantity +
                '}';
    }

}

File: src/main/java/com/hanocybous/ecommercesystem/dto/product/ProductDto.kt
package com.hanocybous.ecommercesystem.dto.product

data class ProductDto(
    val id: Long,
    val name: String,
    val price: Double,
    val category: String,
    val quantity: Int
) {

    init {
        require(name.isNotBlank()) { "Name cannot be blank" }
        require(price >= 0) { "Price cannot be negative" }
        require(category.isNotBlank()) { "Category cannot be blank" }
        require(quantity >= 0) { "Quantity cannot be negative" }
    }

    override fun toString(): String {
        return "ProductDto(id=$id, " +
                "name='$name', " +
                "price=$price, " +
                "category='$category', " +
                "quantity=$quantity)"
    }
}
